src/System.Memory/src/System/MemoryExtensions.cs

```diff
@@ -222,6 +222,50 @@ public static bool StartsWith<T>(this ReadOnlySpan<T> span, ReadOnlySpan<T> valu
             return valueLength <= span.Length && SpanHelpers.SequenceEqual(ref span.DangerousGetPinnableReference(), ref value.DangerousGetPinnableReference(), valueLength);
         }
 
+        /// <summary>
+        /// Determines whether the specified sequence appears at the end of the span.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool EndsWith<T>(this Span<T> span, ReadOnlySpan<T> value)
+            where T : IEquatable<T>
+        {
+            int spanLength = span.Length;
+            int valueLength = value.Length;
+            if (typeof(T) == typeof(byte))
```

Comment:
@ahsonkhan, these checks for typeof(T) == typeof(byte) and the resulting code duplication exists in SequenceEqual, StartsWith, and now EndsWith.  Could the split path instead be pushed down into SpanHelpers.SequenceEqual so that all of the callers that sit on top of it don't need to specialize for byte?

Reply:
It is done intentionally this way to help the JIT generate efficient code for byte specializations. If we push it down, I expect that you will see regressions for byte specializations for small spans.

Reply:
Interesting.  Why does it matter to the JIT whether the branch is at this level or the one it calls?  Something fundamental?  Lack of some optimization that could be applied?

Reply:
One generic methods calling another generic method does not inline for shared generic types. 

If we did not care about that this is going to be slower than what it needs to be for shared generic types, we have more options.

Reply:
> One generic methods calling another generic method does not inline for shared generic types.

Oh, maybe I misunderstood.  I thought you were saying pushing it down would make `Span<byte>` slower, but you're saying it would make `Span<SomeReferenceType>` slower, right?

Reply:
Right. 

There are different options how this check can be pushed down, with different trade-offs.